packager/certcache/certcache.go

```diff
@@ -445,3 +570,171 @@ func (this *CertCache) fetchOCSP(orig []byte, ocspUpdateAfter *time.Time) []byte
 	}
 	return respBytes
 }
+
+// Checks for cert updates every certCheckInterval hours. Never terminates.
+func (this *CertCache) maintainCerts(stop chan struct{}) {
+	// Only make one request per certCheckInterval, to minimize the impact
+	// on servers that are buckling under load.
+	ticker := time.NewTicker(CertCheckInterval)
+
+	for {
+		select {
+		case <-ticker.C:
+			this.updateCertIfNecessary()
+		case <-stop:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// Returns true iff cert cache contains at least 1 cert.
+func (this *CertCache) hasCert() bool {
+	return len(this.certs) > 0 && this.certs[0] != nil
+}
+
+func (this *CertCache) getCert() *x509.Certificate {
+	if !this.hasCert() {
+		return nil
+	}
+	return this.certs[0]
+}
+
+// Set current cert with mutex protection.
+func (this *CertCache) setCerts(certs []*x509.Certificate) {
+	this.certsMu.Lock()
+	defer this.certsMu.Unlock()
+	this.certs = certs
+	this.certName = util.CertName(certs[0])
+
+	err := certloader.WriteCertsToFile(this.certs, this.CertFile)
+	if err != nil {
+		log.Printf("Unable to write certs to file: %s", this.CertFile)
+	}
+}
+
+// Set new cert with mutex protection.
+func (this *CertCache) setNewCerts(certs []*x509.Certificate) {
+	this.renewedCertsMu.Lock()
+	defer this.renewedCertsMu.Unlock()
+	this.renewedCerts = certs
+
+	if this.renewedCerts == nil {
+		err := certloader.RemoveFile(this.NewCertFile)
+		if err != nil {
+			log.Printf("Unable to remove file: %s", this.NewCertFile)
+		}
+		return
+	}
+
+	err := certloader.WriteCertsToFile(this.renewedCerts, this.NewCertFile)
+	if err != nil {
+		log.Printf("Unable to write certs to file: %s", this.NewCertFile)
+	}
+}
+
+// Update the cert in the cache if necessary.
+func (this *CertCache) updateCertIfNecessary() {
+	log.Println("Updating cert if necessary")
+	if this.certFetcher == nil {
+		// Don't request new certs from CA if certFetcher is not set. This means this instance of the amppackager
+		// is not in autorenewcert mode. Just make an attempt at reading the cert saved on disk to see if
+		// another amppackager instance that is in autorenewcert mode actually updated it with a valid cert.
+		log.Println("Certfetcher is not set, skipping cert updates. Checking cert on disk if updated.")
+		this.reloadCertIfExpired()
+		return
+	}
+	d := time.Duration(0)
+	err := errors.New("")
+	if this.hasCert() {
+		d, err = util.GetDurationToExpiry(this.certs[0], time.Now())
+	}
+	if err != nil {
+		if this.renewedCerts != nil {
+			// If renewedCerts is set, copy that over to certs
+			// and set renewedCerts to nil.
+			// TODO(banaag): do the same cert setting dance on disk.
```

Comment:
Is this TODO still accurate? setCerts/setNewCerts seem to write to disk.

Reply:
Removed.